### DISCUSSION TITLE
Regarding the programming languages we're going to use

### DISCUSSION THREAD

**Discussion by cristianMarcial** [uprm-inso4101-2024-2025-s1/semester-project-regiupr#2] [General]
I think that before choosing a programming language for this project, we must take into account the advantages and disadvantages of each language. Two programming languages ​​that are well known and have a standard library to manage ssh-type systems, are Python and Rust. Both have a library manager that allows easy access to standard libraries. Having library manager that supports libraries that support OpenSSH is important because SSH is a communication protocol between servers which it's used by the UPRM enrollment system.

Among Python and Rust there are some advantages and disadvantages:

### **Advantages of Python🐍:**

- It is much easier to use. Its syntax makes it easier than that of other languages ​​and its interpreter is easy to install on the most common operating systems.

- Most, if not all, members of this project are familiar with Python, since we had to use it in the introductory programming course (CIIC3015).

### **Disadvantages of Python:**

- It is an interpreted language; an application made purely in Python CANNOT run on a device that does not have the Python interpreter installed. If you want to create an application independent of an external interpreter, there are two ways to do it; both have disadvantages, and they are the following:

  1. Use software that compiles the Python source code to an executable. The problem with this arises when working with several imported libraries, which the software might have trouble compiling.
  2.  Use a bundler (e.g. _PyInstaller_), which merges all the libraries of the project (including Python) into a single application. Working with a bundler can be difficult and requires additional work that is not required in languages ​​like Rust.

- Python is a language that produces slow applications when "compiled" (Specifically, 71.9 times slower than an application made in C; language in which the puTTY application is made. Source: https://www.researchgate.net/publication/320436353_Energy_efficiency_across_programming_languages_how_do_energy_time_and_memory_relate).

Now, let's talk about Rust.

### **Advantages of Rust🦀:**

- It is a COMPILED language; an application made purely in Rust CAN run on any device with an operating system, even if the Rust compiler has never touched the device. This means that you don't have to waste time creating a bundler or something like that.

- It is a pretty fast language, even faster than C++ in most applications.

### **Disadvantages of Rust:**

- It is much harder to use than other languages. Its syntax is verbose and differs from languages, and its compiler is a bit difficult to install on certain operating systems (e.g. windows), although you can use WSL to run Rust on Windows.

- Most, if not all, members of this project are not familiar with Rust.

Other languages like C++ and Java are even less suitable for this project since they not only present the disadvantages of both Rust and Python, but they don't even have a standardized telnet library and their library manager are a mess to deal with.

### **UPDATE:** 
The issue about compiling Python has been resolved (see https://github.com/uprm-inso4101-2024-2025-s1/semester-project-regiupr/discussions/3#discussion-7113426 on this forum). So, unless someone bring up another negative quality about it, we're going to use Python for this project.

**Comment by christianmedina-dev** [comment]
After evaluating the pros and cons of Python and Rust, I believe Python is the more suitable choice for our project (RegiUPR).

**Team Familiarity and Productivity**: Most team members are already proficient in Python, thanks to our previous course (CIIC3015). This familiarity will allow us to start development immediately. In contrast, learning Rust from scratch would significantly slow down our Milestones progress.

**Ease of Use**: Python's syntax is simple and easy to work with, especially for a team that may have varying levels of experience with programming. Its interpreter is also easy to install on different operating systems. Also Python allows for quick development, which is essential given the project timelines. While Python may have performance limitations, the speed and ease of maintenance often outweigh the need for highly optimized code.

**Library Support**: Python has well-established libraries for managing Telnet-type systems, which are very important for RegiUPR. The availability of these libraries means we won’t need to reinvent the wheel, and can instead focus on implementing the features that matter most.

**Deployment Considerations**: While it’s true that Python apps require an interpreter, modern solutions like PyInstaller or virtual environments can mitigate this issue. These tools allow us to package our application with all its dependencies, ensuring it runs smoothly across different systems.

To conclude given our project requirements, the team's familiarity, and the rapid development advantages, for me Python stands out as the most practical choice. Although Rust offers some performance benefits, the learning curve and complexity make it less suitable for this particular project and the time window available. By choosing Python, we can leverage our existing skills, speed up development, and focus on delivering a high-quality application.

**Reply by MarkoSchuetz** [reply]
As much as I would like to see more modern languages used, I completely agree with the main point and that's familiarity.

**Reply by Ar2691** [reply]
I also agree with the choice of Python, because as previosly mentioned is a familiar high-level programming language, in contrast to rust which might have a steep learning curve.

**Comment by keisanabria** [comment]
It is unfortunate that Python needs an external app to be able to be used for the project that we're developing, but I believe that since Rust "is much harder to use than other languages" overpowers Python's disadvantages since it will cause more problems regarding syntax than the code itself.

**Comment by carlospepin23** [comment]
I agree with the points in favor of Python. While it is true that Python can be slower and may require additional work to create an executable, the time saved using it outweighs the time it would take the team to learn Rust. For this first milestone, it is crucial to make as much progress as possible to get a working prototype; that will be more achievable if we have as many developers ready to code.

**Comment by cristianramos9** [comment]
I agree with using Python as the programming language.
Regardless of the technical advantages and disadvantages of Python and Rust, @cristianMarcial mentioned the two most important points: 

>...
>* Most, if not all, members of this project **are familiar with Python**, since we had to use it in the introductory programming course (CIIC3015).
>...
>* Most, if not all, members of this project **are not familiar with Rust**.
>...

I suggest a poll for choosing from the two languages, closing it at ~9PM on August 30 (or morning of August 31).
Discussing this is important, but I think we should not take too long on deciding.

**Comment by michellebou** [comment]
I agree that it's important to consider the strengths and weaknesses of each programming language before making a decision. Python is known for being user-friendly and familiar to most of us, which makes it easier to work with. However, it can be slower and requires extra steps to create standalone applications.

Rust, while more complex, is faster and can produce independent applications without additional tools. The trade-off is its learning curve, which might slow down development initially.

Both languages have their merits, so the decision should balance ease of use with performance needs.

**Comment by Yamil-Serrano** [comment]
Python is a language that often conceals, it has a wide range of libraries that we can use for this project. I consider it important to select a programming language that many of us already know since not all of us will have an equal learning curve to learn the syntax of a new language, although these are more convenient for us, either due to the availability of a framework or its efficiency or speed.
